### DISCUSSION TITLE
Alternative for setkey for strongswan.

### DISCUSSION THREAD

**Discussion by gowdamanil** [strongswan/strongswan#2221] [Q&A]
Hello @tobiasbrunner 

1. We are currently looking for alternative to racoon and setkey,  we found strongswan for racoon.

But is there an alternative for setkey? Can i use ip xfrm for manipulating spd? 

2. Suppose I want to have kernel policy as below

src local-IP/24 dst 0.0.0.0/0
	dir out priority 2000 
	tmpl src local-IP dst 0.0.0.0
	proto esp reqid 1 mode transport
src local-IP/24 dst 0.0.0.0/0 
	dir in priority 2000 
	tmpl src local-IP dst 0.0.0.0
	proto esp reqid 1 mode transport

What  to use in right and left in swanctl?  Because if i use right = 0.0.0.0/0, then the output of ip xfrm policy show is empty. But when i add specific IP, this is seen?

**Comment by tobiasbrunner** [comment]
> But is there an alternative for setkey?

To do what exactly?

> Can i use ip xfrm for manipulating spd?

Theoretically, yes. But you generally don't have to manipulate the SPD manually. Policies usually are negotiated as traffic selectors.

> Suppose I want to have kernel policy as below

Hm, transport mode with a `/24` and `/0` policy? Seems weird (unless this is intended to be something like the [trap-any scenario](https://strongswan.org/testing/testresults/ikev2/trap-any/)).

**Reply by gowdamanil** [reply]
> To do what exactly?
While I was using racoon, the kernel entries were done by setkey, so now since shifting to strongswan, so need an  alternative.

I will explain my problem in detail.

1. I have configured swanctl.conf as depicted at ending of this whole comment.
2. When i ping from my remote(Here it is client pc, windows). The IPSec is working as expected.
3. But the problem here is, when i ping from windows client, the first packet is seen unencrypted, following all are getting encrypted.
tcpdump -i eth0 host x.x.x.x
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
**11:52:34.522719 IP windowsClient > LinuxMachine: ICMP echo request, id 1, seq 164, length 40
11:52:34.522758 IP LinuxMachine > windowsClient: ICMP echo reply, id 1, seq 164, length 40**
11:52:34.525753 IP windowsClient.isakmp > LinuxMachine.isakmp: isakmp: phase 1 I ident
11:52:34.529338 IP LinuxMachine.isakmp > windowsClient.isakmp: isakmp: phase 1 R ident
11:52:34.531014 IP windowsClient.isakmp > LinuxMachine.isakmp: isakmp: phase 1 I ident
11:52:34.535945 IP LinuxMachine.isakmp > windowsClient.isakmp: isakmp: phase 1 R ident
11:52:34.539458 IP windowsClient.isakmp > LinuxMachine.isakmp: isakmp: phase 1 I ident[E]
11:52:34.539928 IP LinuxMachine.isakmp > windowsClient.isakmp: isakmp: phase 1 R ident[E]
11:52:34.541871 IP windowsClient.isakmp > LinuxMachine.isakmp: isakmp: phase 2/others I oakley-quick[E]
11:52:34.542396 IP LinuxMachine.isakmp > windowsClient.isakmp: isakmp: phase 2/others R oakley-quick[E]
11:52:34.543466 IP windowsClient > LinuxMachine: ESP(spi=0xcdc96e60,seq=0x1), length 76
11:52:34.543642 IP windowsClient.isakmp > LinuxMachine.isakmp: isakmp: phase 2/others I oakley-quick[E]
11:52:35.533107 IP windowsClient > LinuxMachine: ESP(spi=0xcdc96e60,seq=0x2), length 76
11:52:35.533173 IP LinuxMachine > windowsClient: ESP(spi=0x30d2b2a4,seq=0x1), length 76
11:52:36.540290 IP windowsClient > LinuxMachine: ESP(spi=0xcdc96e60,seq=0x3), length 76
11:52:36.540373 IP LinuxMachine > windowsClient: ESP(spi=0x30d2b2a4,seq=0x2), length 76
11:52:37.549649 IP windowsClient > LinuxMachine: ESP(spi=0xcdc96e60,seq=0x4), length 76
11:52:37.549731 IP LinuxMachine > windowsClient: ESP(spi=0x30d2b2a4,seq=0x3), length 76
11:52:39.264955 ARP, Request who-has LinuxMachine (00:43:34:3vf:0u6 (oui Unknown)) tell windowsClient, length 46
11:52:39.265038 ARP, Reply LinuxMachine is-at 00:43:34:3vf:0u6 (oui Unknown), length 28

From above u can see ICMP, but can even they be encrypted? since they are also packets from IPSec enabled client?. For compatibility purpose we are using transport mode ikev1 type of strongswan, we will migrate to ikev2 soon.

Here is swanctl.conf file as told.

connections {
        hostToHost {
                local_addrs=x.x.x.x
                #remote_addrs=0.0.0.0/0
                version=1
                proposals=aes128-sha256-modp2048, 3des-sha1-modp1024,des-sha1-modp1024
                dpd_timeout=120s
                dpd_delay=30s
                children {
                        hostToHost {
                                start_action=start
                                esp_proposals=des-sha1,3des-sha1
                                dpd_action=restart
                                mode = transport
                        }

                }
                local-0 {
                        auth = psk
                }
                remote-0 {
                        auth = psk
                }
        }
}
secrets {
        ike-generic {
                secret = "MySecretKeyHere"
        }
}

Is there any suggestions,  I was thinking if this can be avoided if i can use ip xfrm and add policy like shown in my previous comment but then I saw your comment which was old.https://lists.strongswan.org/pipermail/dev/2012-May/000493.html

When i use start_action = trap, my system is crashing(when client pc do not support IPSec) but no issue with ones that support.

Thanks

**Reply by tobiasbrunner** [reply]
> But the problem here is, when i ping from windows client, the first packet is seen unencrypted, following all are getting encrypted.

That sounds like a Windows problem. If you configure `start_action=start` without remote address that has basically no effect, so the configuration will only be used as responder and as long as Windows does not initiate an SA, no traffic will be protected.

> When i use start_action = trap, my system is crashing

It you want to trigger SAs based on traffic, that's the way to go (as mentioned, `start_action=start` won't work without known remote address). But note that this will block matching inbound traffic (strongSwan currently installs policies in both directions - there is an old patch in the _trap-policies-out-only_ branch that adds a global option to change that).

Also note that when you configure such a wildcard trap policy (transport mode, no traffic selectors, and no remote address), this will match **all** outbound traffic that is not explicitly bypassed (see below) and there will be an attempt to initiate an SA (there is no automatic fallback to sending traffic unencrypted if the negotiation fails). You might want to reconsider that configuration (e.g. change the remote traffic selector to only protect traffic to certain hosts).

Anyway, how exactly is your complete system crashing?

> (when client pc do not support IPSec)

What would you expect to happen with those? Should they be able to communicate without IPsec?  (The latter can only be achieved with strongSwan by manually configuring a [bypass policy](https://docs.strongswan.org/docs/5.9/config/quickstart.html#_passthroughbypass_policies) for such clients, there is no automatism.)

**Reply by gowdamanil** [reply]
Thanks for your suggestion
